### ISSUE TITLE
enable Usage.init_artifact to accept kwargs that are passed to the factory

### ISSUE THREAD

**Issue by gregcaporaso** [qiime2/qiime2#763]
Is there a technical reason why `Usage.init_artifact` doesn't accept kwargs that can be passed through to the factory function? 

If not that would be convenient to enable, to support factory functions like the following:

```python
def dna_artifact_factory(sequence_id: str, sequence: str):
    seq = skbio.DNA(sequence, metadata={id: sequence_id})
    with tempfile.NamedTemporaryFile() as f:
        # write our skbio.DNA object to file in fasta format (the default)
        seq.write(f)
        # reset to the beginning of the file
        f.seek(0)
        # instantiate our format object with the fasta
        ff = SingleRecordDNAFASTAFormat(f.name, mode='r')
        # return the sequence packaged in a "SingleDNASequence" qiime2.Artifact
        return qiime2.Artifact.import_data("SingleDNASequence", ff)

def nw_align_example_1(use):
    seq1 = use.init_artifact('seq1', dna_artifact_factory, "seq1", "ACGGTTAA")
    seq2 = use.init_artifact('seq2', dna_artifact_factory, "seq2", "ATTGTAA")
...
```

I planned to do this while working on usage example content for *Developing with QIIME 2*, but remembered that I need to define a factory function per input as it stands right now.

**Comment by ebolyen** [comment]
Hey @gregcaporaso,

Since there's two sides to this `init_artifact` API, it would be a breaking change and require all drivers to support passing *args and **kwargs (and up through the inheritance chain), which seems like an easy thing to forget to do. Additionally, these drivers would need to wrap that in a closure, as the ultimate call-site may no longer have access to these values (the driver has long since stopped evaluating, and we have a list of variables we could materialize via the factory except for the now-missing arguments). 

I would argue a simpler way to handle this would be to build the closure up front:

```python
def dna_artifact_factory(sequence_id: str, sequence: str):
    def factory():
        seq = skbio.DNA(sequence, metadata={id: sequence_id})
        with tempfile.NamedTemporaryFile() as f:
            # write our skbio.DNA object to file in fasta format (the default)
            seq.write(f)
            # reset to the beginning of the file
            f.seek(0)
            # instantiate our format object with the fasta
            ff = SingleRecordDNAFASTAFormat(f.name, mode='r')
            # return the sequence packaged in a "SingleDNASequence" qiime2.Artifact
            return qiime2.Artifact.import_data("SingleDNASequence", ff)
    return factory

# basically the same, but we are returning the factory with the arguments in a closure now, so that they will always exist wherever the final callsite actually ends up.
def nw_align_example_1(use):
    seq1 = use.init_artifact('seq1', dna_artifact_factory("seq1", "ACGGTTAA"))
    seq2 = use.init_artifact('seq2', dna_artifact_factory("seq2", "ATTGTAA"))
...
```
